Content.Server/StationEvents/Events/MeteorSwarm.cs

```diff
@@ -0,0 +1,126 @@
+        {
+            base.Startup();
+            var robustRandom = IoCManager.Resolve<IRobustRandom>();
+            _waveCounter = robustRandom.Next(MinimumWaves, MaximumWaves);
+        }
+
+        public override void Shutdown()
+        {
+            base.Shutdown();
+            _waveCounter = 0;
+            _cooldown = 0f;
+            EndAfter = float.MaxValue;
+        }
+
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            if (!Started) return;
+
+            if (_waveCounter <= 0)
+            {
+                EndAfter = float.MinValue;
+            }
+            _cooldown -= frameTime;
+
+            if (_cooldown > 0f) return;
+
+            _waveCounter--;
+            var robustRandom = IoCManager.Resolve<IRobustRandom>();
+
+            _cooldown += (MaximumCooldown - MinimumCooldown) * robustRandom.NextFloat() + MinimumCooldown;
+            var entityManager = IoCManager.Resolve<IEntityManager>();
+            var mapManager = IoCManager.Resolve<IMapManager>();
+            var compManager = IoCManager.Resolve<IComponentManager>();
+
+            Box2? playableArea = null;
+            var mapId = EntitySystem.Get<GameTicker>().DefaultMap;
+
+            foreach (var grid in mapManager.GetAllGrids())
+            {
+                if (grid.ParentMapId != mapId || !compManager.TryGetComponent(grid.GridEntityId, out PhysicsComponent? gridBody)) continue;
+                var aabb = gridBody.GetWorldAABB();
+                playableArea ??= aabb;
+                playableArea = playableArea.Value.Union(aabb);
+            }
+
+            if (playableArea == null)
+            {
+                EndAfter = float.MinValue;
+                return;
+            }
+
+            var minimumDistance = (playableArea.Value.TopRight - playableArea.Value.Center).Length + 50f;
+            var maximumDistance = minimumDistance + 100f;
+
+            var center = playableArea.Value.Center;
+
+            for (var i = 0; i < MeteorsPerWave; i++)
+            {
+                var angle = new Angle(robustRandom.NextFloat() * MathF.Tau);
+                var offset = angle.RotateVec(new Vector2((maximumDistance - minimumDistance) * robustRandom.NextFloat() + minimumDistance, 0));
+                var spawnPosition = new MapCoordinates(center + offset, mapId);
+                var meteor = entityManager.SpawnEntity("MeteorLarge", spawnPosition);
+                var physics = compManager.GetComponent<PhysicsComponent>(meteor.Uid);
+                physics.BodyStatus = BodyStatus.InAir;
+                physics.ApplyLinearImpulse(-offset.Normalized * MeteorVelocity * physics.Mass);
+                physics.ApplyAngularImpulse(
+                    // Get a random angular velocity.
+                    physics.Mass * ((MaxAngularVelocity - MinAngularVelocity) * robustRandom.NextFloat() +
+                                    MinAngularVelocity));
```

Comment:
feels like there could be a helper for this, i imagine a lot of stuff would want to do something similar but out of scope

Reply:
Made an issue for it.